src/main.cpp

```diff
@@ -180,857 +219,1155 @@ void forwardIt(const char *line);
+    for (size_t i = 0; i < len; i++) {
+        createAndHandleLine(d[i]);
+    }
+}
 
-  // GET request /scan
-  server.on("/scan", HTTP_GET, [](AsyncWebServerRequest *request)
-            { handleScan(request); });
+static void handleNewClient(void *arg, AsyncClient *client) {
+    clients.push_back(client);
+    client->onData(&handleData, NULL);
+
+#ifdef AISMEMORY
+    std::map<std::string, std::string> ais_messages = getAISMessages();
+    std::map<std::string, std::string>::iterator it;
+
+    for (it = ais_messages.begin(); it != ais_messages.end(); it++) {
+        if (debug_logging) {
+            Serial.print(it->first.c_str());
+            Serial.print(" : ");
+            Serial.println(it->second.c_str());
+        }
+        if (client->space() > it->second.length() && client->canSend()) {
+            client->add(it->second.c_str(), it->second.length());
+            client->send();
+        }
+    }
+#endif
+}
 
-  // GET request /wifi
-  server.on("/wifi", HTTP_GET, [](AsyncWebServerRequest *request)
-            { handleWifi(request); });
+void startTCPNMEAForward(uint16_t port) {
+    if (networkOK) {
+        tcp.reset(new AsyncServer(protocolSettings.tcpPort));
+        tcp.get()->onClient(&handleNewClient, tcp.get());
+        tcp.get()->begin();
+        tcpForwaredOK = true;
+    }
+}
 
-  // GET request /protocol
-  server.on("/protocol", HTTP_GET, [](AsyncWebServerRequest *request)
-            { handleProtocol(request); });
+void stopTCPNMEAForward() {
+    tcpForwaredOK = false;
+    tcp.get()->end();
+    tcp.release();
+}
 
-  // GET request /station
-  server.on("/station", HTTP_GET, [](AsyncWebServerRequest *request)
-            { handleStation(request); });
+void startNMEAForward() {
+    // first stop everything
+    // stopNMEAForward();
 
-  // GET request /system
-  server.on("/system", HTTP_GET, [](AsyncWebServerRequest *request)
-            { handleSystem(request); });
+    if (protocolSettings.tcp) {
+        startTCPNMEAForward(protocolSettings.tcpPort);
+        if (mDNSOK) {
+            MDNS.addService("nmea-0183", "tcp", protocolSettings.tcpPort);
+        }
+    }
 
-  server.onNotFound(notFound);
+    if (protocolSettings.udp) {
+        udpForwaredOK = true;
+        if (mDNSOK) {
+            MDNS.addService("nmea-0183", "udp", protocolSettings.udpPort);
+        }
+    }
+
+    if (protocolSettings.websocket) {
+        websocketOk = true;
+        if (mDNSOK) {
+            MDNS.addService("nmea-0183", "ws", protocolSettings.websocketPort);
+        }
+    }
 }
 
-void startWebServer()
-{
-  if (networkOK)
-  {
-    setupWebServer();
-    server.begin();
-    Serial.println("Webserver started");
-    if (mDNSOK)
-    {
-      MDNS.addService("http", "tcp", 80);
+void stopNMEAForward() {
+    udpForwaredOK = false;
+
+    stopTCPNMEAForward();
+    if (mDNSOK) {
+        mdns_service_remove("_nmea-0183", "_tcp");
+        mdns_service_remove("_nmea-0183", "_udp");
```

Comment:
Is it correct that the advertisement of the websocket is not stopped at this point?

Reply:
The mdns part of the code still needs bit of attention.
https://docs.espressif.com/projects/esp-protocols/mdns/docs/latest/en/index.html

I think also the ' addservice' part is wrong and needs to be fixed

Reply:
Yes I remember some differences between the ESP-IDF and the Arduino example. I found it again.

In `ESPmDNS.cpp` is the `addService(char *name, char *proto, uint16_t port)` which calls `mdns_service_add...`.

I think it could be changed to:
```cpp
mdns_service_add(NULL, "_http", "_tcp", 80, NULL, 0);
mdns_service_add(NULL, "_nmea-0183", "_tcp",  protocolSettings.tcpPort, NULL, 0);
```
etc. directly.

Reply:
At some point in the future there could be support for NMEA OneNet as soon as the final specs are released and it gets some adoption.

Reply:
Yes, but when I tried that, it did not like it as the arduino library used only allows for the `  bool addService(const char *service, const char *proto, uint16_t port)`
But also does not expose the feature that allows for multiple service names with the same protocol.
So for now I renamed the websocket variant `_nmea-0183-ws` to avoid the error.

Will add that commit once I tested some more and also added some of the other fixes